### PR TITLE
Validate resource restriction with jwt decoded token one to one

### PR DESCRIPTION
**Pull request by tzheleznyak**
Validates single resource restriction in one to one way with the decoded token

### CODE REVIEW

File: app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+module Authentication
+  module AuthnJwt
+
+    # This class is responsible for retrieving the correct value from the GCP token
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:4]:
```suggestion
    # This class is responsible for retrieving the correct value from the JWT token
```

File: app/domain/errors.rb [hunk 2]
```diff
@@ -401,6 +401,13 @@ module AuthnJwt
               "   with field in the decoded token that contains the identity",
         code: "CONJ00075E"
       )
+
+      JwtTokenClaimIsMissing = ::Util::TrackableErrorClass.new(
+        msg: "Claim '{0-attribute-name}' is missing from JWT token. " \
+             "Verify that you configured the host with permitted restrictions. " \
+             "In case of Compute Engine token, verify that you requested the token using 'format=full'",
+        code: "CONJ00076E"
```
Comment (sashaCher) [inline app/domain/errors.rb:409]: The error code is already in use by `MissingSecretValue` error.

File: app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    # of the requested attribute.
+    class ValidateRestrictionsOneToOne
+      def initialize(decoded_token:)
+        @decoded_token = decoded_token
+      end
+
+      def valid_restriction?(restriction)
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:11]: Can restriction be `null` or something like this?

File: app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb (hunk 1 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:4]: Done

File: app/domain/errors.rb (hunk 2 above)
Comment (tzheleznyak) [reply app/domain/errors.rb:409]: There are many duplicated in this file . I think that ordering it should be out of scope of this pr

File: app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb (hunk 3 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:11]: Because we do the restriction extraction from the db there is no case it should be null and this case will lead to error when we trying to access name attribute which will make the authentication request to fail

File: app/domain/errors.rb (hunk 2 above)
Comment (sashaCher) [reply app/domain/errors.rb:409]: I'm agree ordering is out of scope for this PR
However I think that choosing a unique code for new error is a part of the PR

**Review by sashaCher — APPROVED** [review]
LGTM

File: app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+module Authentication
+  module AuthnJwt
+
+    # This class is responsible for retrieving the correct value from the JWT token
+    # of the requested attribute.
+    class ValidateRestrictionsOneToOne
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:6]: why is this not a command class?

File: app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      def initialize(decoded_token:)
+        @decoded_token = decoded_token
+      end
+
+      def valid_restriction?(restriction)
+        if @decoded_token.key?(restriction.name.to_sym)
+          return @decoded_token[restriction.name.to_sym] == restriction.value
+        end
+        raise Errors::Authentication::AuthnJwt::JwtTokenClaimIsMissing.new(restriction.name)
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:15]: I would switch this to first raise an error and then check

```suggestion
        unless @decoded_token.key?(restriction.name.to_sym)
          raise Errors::Authentication::AuthnJwt::JwtTokenClaimIsMissing.new(restriction.name)
        end
        return @decoded_token[restriction.name.to_sym] == restriction.value
```

File: app/domain/errors.rb [hunk 6]
```diff
@@ -370,27 +370,27 @@ module AuthnJwt
       InvalidIssuerConfiguration = ::Util::TrackableErrorClass.new(
         msg: "Issuer authenticator configuration is invalid. You should configured as authenticator variables: " \
               "'{0-resource-name}' or one of the following: '{1-resource-name}','{2-resource-name}'",
-        code: "CONJ00070E"
+        code: "CONJ00170E"
```
Comment (orenbm) [inline app/domain/errors.rb:373]: i am guessing these are pushed by mistake?

File: app/domain/errors.rb [hunk 7]
```diff
@@ -399,7 +399,14 @@ module AuthnJwt
...
               "   with field in the decoded token that contains the identity",
-        code: "CONJ00075E"
+        code: "CONJ00175E"
+      )
+
+      JwtTokenClaimIsMissing = ::Util::TrackableErrorClass.new(
+        msg: "Claim '{0-attribute-name}' is missing from JWT token. " \
+             "Verify that you configured the host with permitted restrictions. " \
+             "In case of Compute Engine token, verify that you requested the token using 'format=full'",
+        code: "CONJ00176E"
```
Comment (orenbm) [inline app/domain/errors.rb:409]:
```suggestion
        code: "CONJ00076E"
```
?

File: app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    # This class is responsible for retrieving the correct value from the JWT token
+    # of the requested attribute.
+    class ValidateRestrictionsOneToOne
+      def initialize(decoded_token:)
+        @decoded_token = decoded_token
+      end
+
+      def valid_restriction?(restriction)
+        if @decoded_token.key?(restriction.name.to_sym)
+          return @decoded_token[restriction.name.to_sym] == restriction.value
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:13]: We do not want to raise an exception if it is not equal ?

File: app/domain/errors.rb (hunk 7 above)
Comment (tzheleznyak) [reply app/domain/errors.rb:409]: To avoid duplication

File: app/domain/errors.rb (hunk 2 above)
Comment (tzheleznyak) [reply app/domain/errors.rb]: Changed all error codes of jwt exceptions

File: app/domain/errors.rb (hunk 6 above)
Comment (tzheleznyak) [reply app/domain/errors.rb:373]: No its not
https://github.com/cyberark/conjur/pull/2148#discussion_r627614858

File: app/domain/errors.rb (hunk 7 above)
Comment (tzheleznyak) [reply app/domain/errors.rb:409]: https://github.com/cyberark/conjur/pull/2148#discussion_r627614858

File: app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb (hunk 4 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:6]: Because it is used in the generic resource restriction validatior done by @abrahamko 

Each authenticator gcp for example have file authenticator_request.rb that impliments validation of single restriction (annotation)

For example in gcp it here :
https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_gcp/authentication_request.rb

The usage is here:
https://github.com/cyberark/conjur/blob/e44eaaef953651ebb2458cad6ef69bf6b1a1f040/app/domain/authentication/resource_restrictions/validate_resource_restrictions.rb#L71

File: app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb (hunk 5 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:15]: Fixed

File: app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb (hunk 8 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:13]: The command class using it raises exception if its false
https://github.com/cyberark/conjur/blob/e44eaaef953651ebb2458cad6ef69bf6b1a1f040/app/domain/authentication/resource_restrictions/validate_resource_restrictions.rb#L71

File: app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb (hunk 4 above)
Comment (abrahamko) [inline app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:6]: This class usage is not part of this PR.
It makes this PR very hard to read, since reviewer cannot see how do you use this class.
When a new PR will arrive that uses this code, the reviewer will not have the full context to review its correct usage or comment on the original class, since it is not in that PR.

Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:6]: Yes you are correct.
Just wanted to try to make my pr's smaller then the one with the get identity

Comment (abrahamko) [reply app/domain/authentication/authn_jwt/validate_restrictions_one_to_one.rb:6]: that's good effort, just keep them grouped with some meaning 😄
